File: clients/sdk/frontend/typescript/src/services/messages.ts
import { HTTPClient } from '../client/http';
import { validatePaginationQuery } from '../utils/validation';
import type { MessageCreateRequestType, MessageUpdateRequestType, MessageResponseType, MessagesListResponseType, CreateMessageResponseType, UpdateMessageResponseType, DeleteMessageResponseType, MessageListQueryType } from '../types';

export class MessagesService {
  private httpClient: HTTPClient;

  constructor(httpClient: HTTPClient) {
    this.httpClient = httpClient;
  }

  /**
   * List messages for a thread.
   * @param threadKey thread key
   * @param query optional query parameters (limit, before, after, anchor, sort_by)
   * @param userId optional user id to attach as X-User-ID
   * @param userSignature optional signature to attach as X-User-Signature
   */
  listThreadMessages(threadKey: string, query: MessageListQueryType = {}, userId?: string, userSignature?: string): Promise<MessagesListResponseType> {
    validatePaginationQuery(query);
    const qs = new URLSearchParams();
    if (query.limit !== undefined) qs.set('limit', String(query.limit));
    if (query.before) qs.set('before', query.before);
    if (query.after) qs.set('after', query.after);
    if (query.anchor) qs.set('anchor', query.anchor);
    if (query.sort_by) qs.set('sort_by', query.sort_by);
    return this.httpClient.request(`/frontend/v1/threads/${encodeURIComponent(threadKey)}/messages${qs.toString() ? `?${qs.toString()}` : ''}`, 'GET', undefined, userId, userSignature) as Promise<MessagesListResponseType>;
  }

  /**
   * Create a message within a thread.
   * @param threadKey thread key
   * @param msg message payload
   * @param userId optional user id to send as X-User-ID
   * @param userSignature optional signature to send as X-User-Signature
   */
  createThreadMessage(threadKey: string, msg: MessageCreateRequestType, userId?: string, userSignature?: string): Promise<CreateMessageResponseType> {
    return this.httpClient.request(`/frontend/v1/threads/${encodeURIComponent(threadKey)}/messages`, 'POST', msg, userId, userSignature) as Promise<CreateMessageResponseType>;
  }

  /**
   * Retrieve a message by key within a thread.
   * @param threadKey thread key
   * @param messageKey message key
   * @param userId optional user id to attach as X-User-ID
   * @param userSignature optional signature to attach as X-User-Signature
   */
  getThreadMessage(threadKey: string, messageKey: string, userId?: string, userSignature?: string): Promise<MessageResponseType> {
    return this.httpClient.request(`/frontend/v1/threads/${encodeURIComponent(threadKey)}/messages/${encodeURIComponent(messageKey)}`, 'GET', undefined, userId, userSignature) as Promise<MessageResponseType>;
  }

  /**
   * Update a message within a thread.
   * @param threadKey thread key
   * @param messageKey message key
   * @param msg message payload
   * @param userId optional user id to attach as X-User-ID
   * @param userSignature optional signature to attach as X-User-Signature
   */
  updateThreadMessage(threadKey: string, messageKey: string, msg: MessageUpdateRequestType, userId?: string, userSignature?: string): Promise<UpdateMessageResponseType> {
    return this.httpClient.request(`/frontend/v1/threads/${encodeURIComponent(threadKey)}/messages/${encodeURIComponent(messageKey)}`, 'PUT', msg, userId, userSignature) as Promise<UpdateMessageResponseType>;
  }

  /**
   * Soft-delete a message within a thread.
   * @param threadKey thread key
   * @param messageKey message key
   * @param userId optional user id to attach as X-User-ID
   * @param userSignature optional signature to attach as X-User-Signature
   */
  deleteThreadMessage(threadKey: string, messageKey: string, userId?: string, userSignature?: string): Promise<DeleteMessageResponseType> {
    return this.httpClient.request(`/frontend/v1/threads/${encodeURIComponent(threadKey)}/messages/${encodeURIComponent(messageKey)}`, 'DELETE', undefined, userId, userSignature) as Promise<DeleteMessageResponseType>;
  }
}

File: clients/sdk/backend/python/src/progressdb/services/messages.py
"""Messages service for ProgressDB Python SDK."""

from typing import TYPE_CHECKING, Optional
from urllib.parse import urlencode

from ..types.queries import MessageListQueryType
from ..types.responses import (
    CreateMessageResponseType,
    DeleteMessageResponseType,
    MessageResponseType,
    MessagesListResponseType,
    UpdateMessageResponseType,
)
from ..types.message import MessageCreateRequestType, MessageUpdateRequestType
from ..utils.validation import validate_pagination_query

if TYPE_CHECKING:
    from ..client.http import HTTPClient


class MessagesService:
    """Service for message operations."""

    def __init__(self, http_client: "HTTPClient") -> None:
        """Initialize messages service.
        
        Args:
            http_client: HTTP client instance
        """
        self.http_client = http_client

    async def list_thread_messages(
        self,
        thread_key: str,
        query: MessageListQueryType = {},
        user_id: Optional[str] = None,
        user_signature: Optional[str] = None,
    ) -> MessagesListResponseType:
        """List messages for a thread.
        
        Args:
            thread_key: Thread key
            query: Optional query parameters (limit, before, after, anchor, sort_by)
            user_id: Optional user id to attach as X-User-ID
            user_signature: Optional signature to attach as X-User-Signature
            
        Returns:
            List of messages
        """
        validate_pagination_query(query)
        
        qs_params = {}
        if query.get("limit") is not None:
            qs_params["limit"] = str(query["limit"])
        if query.get("before"):
            qs_params["before"] = query["before"]
        if query.get("after"):
            qs_params["after"] = query["after"]
        if query.get("anchor"):
            qs_params["anchor"] = query["anchor"]
        if query.get("sort_by"):
            qs_params["sort_by"] = query["sort_by"]
            
        query_string = urlencode(qs_params)
        encoded_thread_key = self._encode_url_component(thread_key)
        url = f"/frontend/v1/threads/{encoded_thread_key}/messages{f'?{query_string}' if query_string else ''}"
        
        return await self.http_client.request(url, "GET", None, user_id, user_signature)

    async def create_thread_message(
        self,
        thread_key: str,
        msg: MessageCreateRequestType,
        user_id: Optional[str] = None,
        user_signature: Optional[str] = None,
    ) -> CreateMessageResponseType:
        """Create a message within a thread.
        
        Args:
            thread_key: Thread key
            msg: Message payload
            user_id: Optional user id to send as X-User-ID
            user_signature: Optional signature to send as X-User-Signature
            
        Returns:
            Created message response
        """
        encoded_thread_key = self._encode_url_component(thread_key)
        return await self.http_client.request(
            f"/frontend/v1/threads/{encoded_thread_key}/messages",
            "POST",
            msg,
            user_id,
            user_signature,
        )

    async def get_thread_message(
        self,
        thread_key: str,
        message_key: str,
        user_id: Optional[str] = None,
        user_signature: Optional[str] = None,
    ) -> MessageResponseType:
        """Retrieve a message by key within a thread.
        
        Args:
            thread_key: Thread key
            message_key: Message key
            user_id: Optional user id to attach as X-User-ID
            user_signature: Optional signature to attach as X-User-Signature
            
        Returns:
            Message data
        """
        encoded_thread_key = self._encode_url_component(thread_key)
        encoded_message_key = self._encode_url_component(message_key)
        return await self.http_client.request(
            f"/frontend/v1/threads/{encoded_thread_key}/messages/{encoded_message_key}",
            "GET",
            None,
            user_id,
            user_signature,
        )

    async def update_thread_message(
        self,
        thread_key: str,
        message_key: str,
        msg: MessageUpdateRequestType,
        user_id: Optional[str] = None,
        user_signature: Optional[str] = None,
    ) -> UpdateMessageResponseType:
        """Update a message within a thread.
        
        Args:
            thread_key: Thread key
            message_key: Message key
            msg: Message payload
            user_id: Optional user id to attach as X-User-ID
            user_signature: Optional signature to attach as X-User-Signature
            
        Returns:
            Updated message response
        """
        encoded_thread_key = self._encode_url_component(thread_key)
        encoded_message_key = self._encode_url_component(message_key)
        return await self.http_client.request(
            f"/frontend/v1/threads/{encoded_thread_key}/messages/{encoded_message_key}",
            "PUT",
            msg,
            user_id,
            user_signature,
        )

    async def delete_thread_message(
        self,
        thread_key: str,
        message_key: str,
        user_id: Optional[str] = None,
        user_signature: Optional[str] = None,
    ) -> DeleteMessageResponseType:
        """Soft-delete a message within a thread.
        
        Args:
            thread_key: Thread key
            message_key: Message key
            user_id: Optional user id to attach as X-User-ID
            user_signature: Optional signature to attach as X-User-Signature
            
        Returns:
            Delete response
        """
        encoded_thread_key = self._encode_url_component(thread_key)
        encoded_message_key = self._encode_url_component(message_key)
        return await self.http_client.request(
            f"/frontend/v1/threads/{encoded_thread_key}/messages/{encoded_message_key}",
            "DELETE",
            None,
            user_id,
            user_signature,
        )

    def _encode_url_component(self, component: str) -> str:
        """Encode URL component (similar to encodeURIComponent in JS).
        
        Args:
            component: String to encode
            
        Returns:
            URL-encoded string
        """
        # Python's quote is similar to encodeURIComponent
        from urllib.parse import quote
        return quote(component, safe='')